HAL/keymaster/4.1/JavacardKeymaster4Device.cpp

```diff
@@ -1071,78 +1025,161 @@ Return<void> JavacardKeymaster4Device::begin(KeyPurpose purpose, const hidl_vec<
-                            !cborConverter_.getUint64(item, 2, operationHandle)) {
+                    if (!cborConverter_.getKeyParameters(item, 1, outParams) ||
+                        !cborConverter_.getUint64(item, 2, operationHandle)) {
                         errorCode = ErrorCode::UNKNOWN_ERROR;
                         outParams.setToExternal(nullptr, 0);
                         operationHandle = 0;
-                        LOG(ERROR) << "INS_BEGIN_OPERATION_CMD: error in converting cbor data, status: " << (int32_t) errorCode;
+                        LOG(ERROR) << "INS_BEGIN_OPERATION_CMD: error in converting cbor "
+                                   "data, status: "
+                                   << (int32_t)errorCode;
                     } else {
                         /* Store the operationInfo */
-                        oprCtx_->setOperationInfo(operationHandle, purpose, param.f.algorithm, inParams);
+                        oprCtx_->setOperationInfo(operationHandle, purpose,
+                                                  param.f.algorithm, inParams);
                     }
                 }
             }
         } else {
-            LOG(ERROR) << "INS_BEGIN_OPERATION_CMD couldn't find algorithm tag: " << (int32_t)Tag::ALGORITHM;
+            LOG(ERROR) << "INS_BEGIN_OPERATION_CMD couldn't find algorithm tag: "
+                       << (int32_t)Tag::ALGORITHM;
         }
     } else {
-        LOG(ERROR) << "INS_BEGIN_OPERATION_CMD error in getKeyCharacteristics status: " << (int32_t) errorCode;
+        LOG(ERROR)
+                << "INS_BEGIN_OPERATION_CMD error in getKeyCharacteristics status: "
+                << (int32_t)errorCode;
+    }
+    return errorCode;
+}
+
+ErrorCode JavacardKeymaster4Device::handleBeginOperation(
+    KeyPurpose purpose, const hidl_vec<uint8_t>& keyBlob,
+    const hidl_vec<KeyParameter>& inParams, const HardwareAuthToken& authToken,
+    hidl_vec<KeyParameter>& outParams, uint64_t& operationHandle,
+    OperationType& operType) {
+    ErrorCode errorCode = ErrorCode::UNKNOWN_ERROR;
+    if (operType == OperationType::PUBLIC_OPERATION) {
+        errorCode = handleBeginPublicKeyOperation(purpose, keyBlob, inParams,
+                                                  outParams, operationHandle);
+
+        // For Symmetric operations handleBeginPublicKeyOperation function
+        // returns INCOMPATIBLE_ALGORITHM error. Based on this error
+        // condition it fallbacks to private key operation.
+        if (errorCode == ErrorCode::INCOMPATIBLE_ALGORITHM) {
+            operType = OperationType::PRIVATE_OPERATION;
+        }
+    }
+
+    if (operType == OperationType::PRIVATE_OPERATION) {
+        errorCode = handleBeginPrivateKeyOperation(
+            purpose, keyBlob, inParams, authToken, outParams, operationHandle);
+    }
+
+    // check if operation handle exists.
+    if (ErrorCode::OK == errorCode && isOperationHandleExists(operationHandle)) {
```

Comment:
@BKSSMVenkateswarlu  can you move this part of code to ::begin itself ? It's confusing that you do this check here and again immediately after returning from this function.

Reply:
Modified the code as per the review comments.